e2e/e2e_test.go

```diff
@@ -0,0 +1,151 @@
+package e2e_test
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	bftclient "github.com/cometbft/cometbft/rpc/client/http"
+	opbindings "github.com/ethereum-optimism/optimism/op-bindings/bindings"
+	"github.com/ethereum-optimism/optimism/op-node/bindings"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer/e2e"
+	"github.com/polymerdao/monomer/e2e/url"
+	"github.com/polymerdao/monomer/environment"
+	"github.com/polymerdao/monomer/opdevnet"
+	"github.com/stretchr/testify/require"
+)
+
+var e2eTests = []struct {
+	name string
+	run  func(t *testing.T, stack *e2e.StackConfig)
+}{
+	{
+		name: "ETH L1 Deposits and L2 Withdrawals",
+		run:  ethRollupFlow,
+	},
+	{
+		name: "ERC-20 L1 Deposits",
+		run:  erc20RollupFlow,
+	},
+	{
+		name: "AttributesTX",
+		run:  containsAttributesTx,
+	},
+	{
+		name: "No Rollbacks",
+		run:  checkForRollbacks,
+	},
+}
+
+func TestE2E(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping e2e tests in short mode")
+	}
+
+	env := environment.New()
+	defer func() {
+		require.NoError(t, env.Close())
+	}()
+	artifactsDir, err := filepath.Abs("artifacts")
+	require.NoError(t, err)
+
+	if err := os.Mkdir(artifactsDir, 0o755); !errors.Is(err, os.ErrExist) {
+		require.NoError(t, err)
+	}
+
+	stdoutFile, err := os.OpenFile(filepath.Join(artifactsDir, "stdout"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o666)
+	require.NoError(t, err)
+	env.DeferErr("close stdout file", stdoutFile.Close)
+	stdout := os.Stdout
+	os.Stdout = stdoutFile
+	defer func() {
+		os.Stdout = stdout
+	}()
+	stderrFile, err := os.OpenFile(filepath.Join(artifactsDir, "stderr"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o666)
+	require.NoError(t, err)
+	env.DeferErr("close stderr file", stderrFile.Close)
+	stderr := os.Stderr
+	os.Stderr = stderrFile
+	defer func() {
+		os.Stderr = stderr
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	require.NoError(t, e2e.Run(ctx, env, t.TempDir()))
+
+	for _, test := range e2eTests {
+		t.Run(test.name, func(t *testing.T) {
+			test.run(t, newStackConfig(t))
+		})
+	}
+}
+
+func newStackConfig(t *testing.T) *e2e.StackConfig {
+	secrets, err := opdevnet.DefaultMnemonicConfig.Secrets()
+	require.NoError(t, err)
+
+	{
+		engineurl, err := url.ParseString("ws://127.0.0.1:9000")
+		require.NoError(t, err)
+		engineurl.IsReachable(context.Background())
+	}
+	monomerRPCClient, err := rpc.Dial("ws://127.0.0.1:9000")
+	require.NoError(t, err)
+	monomerClient := e2e.NewMonomerClient(monomerRPCClient)
+
+	bftClient, err := bftclient.New("tcp://127.0.0.1:26657", "/websocket")
+	require.NoError(t, err)
+	require.NoError(t, bftClient.Start())
+
+	{
+		engineurl, err := url.ParseString("ws://127.0.0.1:9001")
+		require.NoError(t, err)
+		engineurl.IsReachable(context.Background())
```

Comment:
Do we need to assert that the engineurl is reachable in both of these places?

`require.True(t, engineurl.IsReachable(context.Background()))`

Reply:
Ah yeah forgot to refactor this. Thanks